packages/mobile-native-mcp-server/src/tools/magi/prd/magi-prd-gap-requirements/tool.ts

```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright (c) 2025, salesforce.com, inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: MIT
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
+ */
+
+import { McpServer } from '@modelcontextprotocol/sdk/server/mcp.js';
+import { Logger } from '../../../../logging/logger.js';
+import { GAP_REQUIREMENTS_TOOL, GapRequirementsInput } from './metadata.js';
+import { PRDAbstractWorkflowTool } from '../../../base/prdAbstractWorkflowTool.js';
+
+/**
+ * Tool for generating functional requirements based on identified gaps.
+ */
+export class MagiGapRequirementsTool extends PRDAbstractWorkflowTool<typeof GAP_REQUIREMENTS_TOOL> {
+  constructor(server: McpServer, logger?: Logger) {
+    super(server, GAP_REQUIREMENTS_TOOL, 'GapRequirementsTool', logger);
+  }
+
+  public handleRequest = async (input: GapRequirementsInput) => {
+    const guidance = this.generateGapRequirementsGuidance(input);
+    const finalOutput = this.finalizeWorkflowToolOutput(guidance, input.workflowStateData);
+    return finalOutput;
+  };
+
+  private generateGapRequirementsGuidance(input: GapRequirementsInput) {
+    const gapsList = input.identifiedGaps
+      .map(
+        (gap, index) => `${index + 1}. **${gap.title}** (${gap.severity} severity)
```

Comment:
This doesn't have to be now, but generally speaking when data manipulation logic gets sufficiently tangled inline with the prompt template it's populating, I prefer to pull out the data manipulation logic into its own method, and encapsulate that work away from the prompt template.

Honestly at some point we probably need to move to a strict prompt template model, where the prompt template text is wholly static, and we replace inline keys with passed-in string values.

Reply:
It's a good call-out that "where the prompt template text is wholly static" has some nuance in use cases like this, where we're actually composing different parts of the template and required output format(s). Discussed some possible approaches offline.